File: 22-generate-parentheses/solution.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        self.add_paran(result, '', n, left=n, right=n)
        return result

    def add_paran(self, result, s_str, n,  left, right):
        if left == 0 and right == 0:
            result.append(s_str)
            return

        if left > 0:
            self.add_paran(result, s_str + '(', n, left - 1, right)
        if n-left > n-right:
            self.add_paran(result, s_str + ')', n, left, right - 1)

File: 22-generate-parentheses/solution.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        if (n == 0) return {};
        vector<string> res;
        string track;
        backtrack(n, n, track, res);
        return res;
    }

    void backtrack(int left, int right, string& track, vector<string>& res) {
        if (right < left) return;
        if (left < 0 || right < 0) return;
        if (left == 0 && right == 0) {
            res.push_back(track);
            return;
        }
        track.push_back('(');
        backtrack(left-1, right, track, res);
        track.pop_back();

        track.push_back(')');
        backtrack(left, right-1, track, res);
        track.pop_back();
    }
};
